components/esp_modem/src/esp_modem_dce.cpp

```diff
@@ -191,4 +262,114 @@ modem_mode DCE_Mode::get()
     return mode;
 }
 
+modem_mode DCE_Mode::guess(DTE *dte, bool with_cmux)
+{
+    Scoped<DTE> lock(*dte);
+    return guess_unsafe(dte, with_cmux);
+}
+
+/**
+ * This namespace contains probe packets and expected replies on 3 different protocols,
+ * the modem device could use (as well as timeouts and mode ids for synchronisation)
+ */
+namespace probe {
+
+namespace ppp {
+// Test that we're in the PPP mode by sending an LCP protocol echo request and expecting LCP echo reply
+const uint8_t lcp_echo_request[] = {0x7e, 0xff, 0x03, 0xc0, 0x21, 0x09, 0x01, 0x00, 0x08, 0x99, 0xd1, 0x35, 0xc1, 0x8e, 0x2c, 0x7e };
+const uint8_t lcp_echo_reply_head[] = {0x7e, 0xff, 0x7d, 0x23, 0xc0};
+const size_t mode = 1 << 0;
+const int timeout = 200;
+}
+
+namespace cmd {
+// For command mode, we just send a simple AT command
+const char at[] = "\r\nAT\r\n";
+size_t max_at_reply = 16; // account for some whitespaces and/or CMUX encapsulation
+const char reply[] = { 'O', 'K' };
+const int mode = 1 << 1;
+const int timeout = 500;
+}
+
+namespace cmux {
+// For CMUX mode, we send an SABM on control terminal (0)
+const uint8_t sabm0_reqest[] = {0xf9, 0x03, 0x3f, 0x01, 0x1c, 0xf9};
+const uint8_t sabm0_reply[] = {0xf9, 0x03, 0x73, 0x01};
+const int mode = 1 << 0;
+const int timeout = 200;
+}
+};
+
+modem_mode DCE_Mode::guess_unsafe(DTE *dte, bool with_cmux)
+{
+    // placeholder for reply and its size, since it could come in pieces, and we have to cache
+    // this is passed to the lambda as references.
+    // must make sure the lambda is cleared before exiting this function (done by dte->on_read(nullptr))
+    uint8_t reply[std::max(probe::cmd::max_at_reply, std::max(sizeof(probe::ppp::lcp_echo_request), sizeof(probe::cmux::sabm0_reply)))];
+    size_t reply_pos = 0;
+    auto signal = std::make_shared<SignalGroup>();
+    std::weak_ptr<SignalGroup> weak_signal = signal;
+    dte->on_read([weak_signal, with_cmux, &reply, &reply_pos](uint8_t *data, size_t len) {
+        // storing the response in the `reply` array and de-fragmenting
+        if (reply_pos >= sizeof(reply)) {
+            return command_result::TIMEOUT;
+        }
+        auto reply_size = std::min((size_t)sizeof(reply) - reply_pos, len);
+        ::memcpy(reply + reply_pos, data, reply_size);
+        reply_pos += reply_size;
+        ESP_LOG_BUFFER_HEXDUMP("esp-modem: guess mode data:", reply, reply_pos, ESP_LOG_DEBUG);
+
+        // Check whether the response resembles the "golden" reply (for these 3 protocols)
+        if (reply_pos >= sizeof(probe::ppp::lcp_echo_reply_head)) {
+            // check for initial 2 bytes
+            uint8_t *ptr = static_cast<uint8_t *>(memmem(reply, reply_pos, probe::ppp::lcp_echo_reply_head, 2));
```

Comment:
Build a `string_view` from the received buffer and use provided algorithms might lead to simpler code here.

Reply:
any idea how to do it without reinterpret cast?
```cpp
    std::string_view reply_view(reinterpret_cast<const char*>(reply), reply_size);
    
    // Create a string_view for the bytes to compare against
    std::string_view head_view(reinterpret_cast<const char*>(ppp::lcp_echo_reply_head.data()), 2);

    // Check if the reply starts with the first 2 bytes of lcp_echo_reply_head
    return reply_view.starts_with(head_view);
```

Reply:
The thing is that I'm actually testing if the current protocol is binary or ascii, so I think the best option is to just use the old good memmem()...